Add slang support to the framework

SaschaWillems opened the pull request:
## Description

This PR adds initial support to slang shaders to the framework, the same way our framework supports offline compiled glsl and hlsl shaders. 

* When running a sample one can select slang shaders using the shading language argument: `sample profiles --shading-language slang`
* If slang shaders are present for said sample, they are loaded (offline compiled SPIR-V)
* If slang shaders are present for samples, they're added to the project file
* Slang shaders are handled via CMake just like HLSL and GLSL shades
* This also means Slang shaders are off-line compiled similar to HLSL (requires an installation of slangc, e.g. from the LunarG SDK)
* Slang shaders are added via the CMakeLists.txt of a sample using `SHADER_FILES_SLANG`

**Note:** While slang supports mutltiple shaders in one entry point, adding that would be a big effort, and something we can tackle in another PR. So for this to work, samples still need to separate slang shaders just like we do for glsl and hlsl.

To make it easier to test, this PR adds slang shaders (almost carbon copy of hlsl, so there might be warnings) to the profiles sample.

Fixes #1234

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

 If this PR contains framework changes:
 - [x] I did a full batch run using the `batch` command line argument to make sure all samples still work properly

SaschaWillems committed:
Started adding support for compiling and using slang shaders

SaschaWillems committed:
Started adding support for compiling and using slang shaders

SaschaWillems committed:
Merge branch 'main' into slang_support

SaschaWillems committed:
Merge branch 'main' into slang_support

SaschaWillems committed:
Update copyrights

SaschaWillems committed:
Fix formatting

SaschaWillems committed:
Update copyright

asuessenbach commented on app/CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2019-2024, Arm Limited and Contributors
+# Copyright (c) 2019-2025, Arm Limited and Contributors
```
Changed nothing but the copyright here?

SaschaWillems replied:
Yeah, that was a desparate try to fix CI. Will revert.

asuessenbach commented on bldsys/cmake/sample_helper.cmake:
```diff
@@ -154,11 +163,15 @@ function(add_project)
         set_source_files_properties(SOURCE ${SHADERS_HLSL} PROPERTIES VS_SETTINGS "ExcludedFromBuild=true")        
     endif()
+    if (TARGET_SHADERS_SLANG)
+        message(STATUS ${TARGET_SHADERS_SLANG})
```
Is this message still important?

asuessenbach commented on framework/api_vulkan_sample.cpp:
```diff
@@ -454,6 +454,14 @@ VkPipelineShaderStageCreateInfo ApiVulkanSample::load_shader(const std::string &
 		shader_extension = ".spv";
 	}
 
+	if (get_shading_language() == vkb::ShadingLanguage::SLANG)
```
Maybe make it an `else if`

SaschaWillems replied:
Good point. I'll rework that part.

asuessenbach commented on framework/api_vulkan_sample.cpp:
```diff
@@ -454,6 +454,14 @@ VkPipelineShaderStageCreateInfo ApiVulkanSample::load_shader(const std::string &
+		src_language     = vkb::ShaderSourceLanguage::SPV;
+		shader_extension = ".spv";
+	}
+
```
As part of those two cases are identical, maybe it's better to have something like
```
auto shading_language = get_shading_language();
if ((shading_language == vkb::ShadingLanguage::HSLS) || (shading_language == vkb::ShadingLanguage::SLANG))
{
    shader_folder = (shading_language == HLSL) ? "hlsl" : "slang";
    ...
}
```

SaschaWillems replied:
I'd prefer to have them separate for better readability and future flexibility. I'll rework the if blocks though as you hinted above.

asuessenbach replied:
Regarding the better readability: I find it better readable, if there are no identical code sequences for the different cases. You (as a human) have to carefully scan the two code sequences, just to realize they actually are identical. In such a case, I would always ask if that's correct

asuessenbach commented on framework/hpp_api_vulkan_sample.cpp:
```diff
@@ -427,6 +427,14 @@ vk::PipelineShaderStageCreateInfo HPPApiVulkanSample::load_shader(const std::str
+		src_language     = vkb::ShaderSourceLanguage::SPV;
+		shader_extension = ".spv";
+	}
+
```
Same comment as in api_vulkan_sample.cpp.

asuessenbach reviewed (commented):
Just some minor issues (besides the missing slang shaders, of course).

asuessenbach commented:
... and in the description, you probably meant to say `sample hello_triangle --shading-language slang` (Note the `--` in front of the `shading-language`)

And maybe more important here: none of the `hello_triangle` flavours care about the shading language, they just use GLSL.

SaschaWillems committed:
Revert unnecessary changes

SaschaWillems commented:
> ... and in the description, you probably meant to say `sample hello_triangle --shading-language slang` (Note the `--` in front of the `shading-language`)
> 
> And maybe more important here: none of the `hello_triangle` flavours care about the shading language, they just use GLSL.

Updated the description. Though it would be nice if at some point the hello triangle samples could also support different shading languages ;)

SaschaWillems committed:
Addd slang shader offline compilation

Add sample shaders to profiles sample for testing this PR

SaschaWillems commented:
Added offline compilation and added slang shaders to the profiles sample (almost carbon copies from hlsl, so there might be some slang warning).

So people can actually test this.

SaschaWillems committed:
Update copyright

SRSaunders commented:
The iOS build may be failing due to an updated Xcode dev environment on the CI machine.  I suspect this can be corrected by changing `-DCMAKE_OSX_DEPLOYMENT_TARGET=13.0` to `-DCMAKE_OSX_DEPLOYMENT_TARGET=14.0` (for Xcode 15.3 or 15.4 used in this case) or `-DCMAKE_OSX_DEPLOYMENT_TARGET=14.5` (for XCode 16.x).  I don't know how the CI setups match to Xcode versions, so I can't say absolutely.  This change would have to be made in the iOS CI build setup in build.yml.

Note this build is also exhibiting the same cmake 3.31 issue solved in #1296.  The iOS CI is likely using an updated dev environment vs. before.

SaschaWillems commented:
I don't quite understand. Another build just a few minutes ago worked?

SaschaWillems committed:
Use combined image sampler

SRSaunders commented:
> I don't quite understand. Another build just a few minutes ago worked?

@SaschaWillems you are correct, I saw the Xcode error 65 and jumped to conclusions - my bad.

The iOS CI build failure is actually caused by Vulkan not being found in the cache, and then falling back to downloading and installing **SDK 1.3.290** (as per the logic in build.yml):

![Screenshot 2025-03-17 at 7 48 36 PM](https://github.com/user-attachments/assets/3af9f235-8ec9-4a39-adff-ee03cd905f64)


The actual failure (which I can duplicate in my dev environment using **SDK 1.3.290**) is :

_The following build commands failed:
	PhaseScriptExecution Generate\ samples/tooling/profiles/shader-slang-spv/profiles.vert.spv /Users/steve/XcodeProjects/Vulkan-Samples/build/ios/build/vulkan_samples.build/Release-iphoneos/profiles-vert-slang.build/Script-3A321516860A670084B67032.sh (in target 'profiles-vert-slang' from project 'vulkan_samples')
(1 failure)_

To correct this, the fallback for iOS Vulkan installs should be brought up to date.  I tested with **SDK 1.4.304** and this works fine with the iOS build commands.  The **build.yml** file needs to be updated with a more modern Vulkan SDK fallback if not found in the cache.  See below...

```
      - name: retrieve VulkanSDK
        if: steps.VulkanSDK.outputs.cache-hit != 'true'
        run: |
          wget https://sdk.lunarg.com/sdk/download/1.3.290.0/mac/vulkansdk-macos-1.3.290.0.dmg
          sudo hdiutil attach vulkansdk-macos-1.3.290.0.dmg
          sudo /Volumes/vulkansdk-macos-1.3.290.0/InstallVulkan.app/Contents/MacOS/InstallVulkan --root ~/VulkanSDK --accept-licenses --default-answer --confirm-command install com.lunarg.vulkan.ios
          sudo hdiutil detach /Volumes/vulkansdk-macos-1.3.290.0
```

SaschaWillems commented:
Is there a way to see more of that error? It tried to add slang in such a way that builds don't fail if the skang compiler isn't available (e.g. with older SDKs). Might have made an error though. Older SDK versions don't have slang, and I try to check for it's presence.

And thanks for looking into this 👍🏻

SRSaunders commented:
> Is there a way to see more of that error?

I ran the build interactively in Xcode and here is the full error...

_PhaseScriptExecution Generate\ samples/tooling/profiles/shader-slang-spv/profiles.vert.spv /Users/steve/XcodeProjects/Vulkan-Samples/build/ios-sdk290/build/vulkan_samples.build/Debug-iphoneos/profiles-vert-slang.build/Script-3A321516860A670084B67032.sh (in target 'profiles-vert-slang' from project 'vulkan_samples')
    cd /Users/steve/XcodeProjects/Vulkan-Samples
    /bin/sh -c /Users/steve/XcodeProjects/Vulkan-Samples/build/ios-sdk290/build/vulkan_samples.build/Debug-iphoneos/profiles-vert-slang.build/Script-3A321516860A670084B67032.sh_

_/Users/steve/XcodeProjects/Vulkan-Samples/build/ios-sdk290/build/vulkan_samples.build/Debug-iphoneos/profiles-vert-slang.build/Script-3A321516860A670084B67032.sh: line 5: Vulkan_slang_EXECUTABLE-NOTFOUND: command not found
Command PhaseScriptExecution failed with a nonzero exit code_

And from cmake:

_-- Found Vulkan: /Users/steve/VulkanSDK/1.3.290.0/iOS/lib/vulkan.framework (found version "1.3.290") found components: glslc glslangValidator
-- Found DirectX Shader Compiler under /Users/steve/VulkanSDK/1.3.290.0/iOS/Bin/dxc
-- Couldn't find slang Shader Compiler executable, make sure it is present in Vulkan SDK or add it manually via Vulkan_slang_EXECUTABLE cmake variable. Slang shaders won't be compiled._

SaschaWillems commented:
Thanks a lot 👍🏻 

Slang shaders should only compile if slangc is found, so my code needs some fixing. Will take a look at that.

Edit: Fixed. I was checking for the wrong shader compiler.

SaschaWillems committed:
Correctly check for slang compiler presence

SRSaunders commented:
@SaschaWillems note that depending on the CI machine you encounter (x86-based or arm-based), you may also run into problem #1294 for iOS CI builds.

If you do, then you will need to merge either PR #1295 or #1300 to fix.

SaschaWillems commented:
I only use the CI machine (?) from our CI setup here at github. So it should be fine now. I want to keep other changes out if this PR if possible.

SRSaunders commented:
No problem, this just a heads up if the iOS CI build fails again...  Hopefully they are arm-based and iOS CI will not fail.

SaschaWillems committed:
Use else if for shading language distinction

gpx1000 commented on shaders/profiles/slang/profiles.frag.slang:
```diff
@@ -0,0 +1,32 @@
+ * limitations under the License.
+ */
+
+[[vk::binding(1, 0)]]
```
Being explicit with the binding is likely not needed.  This is one of the cool differences between SLang and HLSL.

SaschaWillems replied:
Yeah, as noted in the description I only added those shaders so people can actually test the PR. They're mostly carbon copies of the HLSL ones.

gpx1000 commented on shaders/profiles/slang/profiles.frag.slang:
```diff
@@ -0,0 +1,32 @@
+{
+    float4 Pos : SV_POSITION;
+    [[vk::location(0)]] float2 UV : TEXCOORD0;
+    [[vk::location(1)]] nointerpolation int TexIndex : TEXCOORD1;
```
explicit location designation here is also likely not needed.

SaschaWillems replied:
See comment above.

gpx1000 commented on shaders/profiles/slang/profiles.vert.slang:
```diff
@@ -0,0 +1,49 @@
+{
+    [[vk::location(0)]] float3 Pos : POSITION0;
+    [[vk::location(1)]] float2 UV : TEXCOORD0;
+    [[vk::location(2)]] int TexIndex : TEXCOORD1;
```
You likely can replace this with:
struct VSInput {
float3 Pos;
flaot2 UV;
int TexIndex;
}

SaschaWillems replied:
See comment above.

gpx1000 commented on shaders/profiles/slang/profiles.vert.slang:
```diff
@@ -0,0 +1,49 @@
+    float4x4 view;
+};
+[[vk::binding(0, 0)]]
+ConstantBuffer<UBO> ubo : register(b0, space0);
```
don't need either the binding nor the register.

SaschaWillems replied:
See comment above.

gpx1000 commented on bldsys/cmake/sample_helper.cmake:
```diff
@@ -253,4 +266,38 @@ endif()
+            set(OUTPUT_FILE ${OUTPUT_DIR}/${bare_name}.${extension}.spv)
+            file(MAKE_DIRECTORY ${OUTPUT_DIR})
+
+            set(SLANG_PROFILE "glsl_460")
```
Why limit to glsl?  Can just emit directly with spirv_1_4

SaschaWillems replied:
Thx, will take a look at that. I took that from glslangs official documentation ;)

gpx1000 commented on bldsys/cmake/sample_helper.cmake:
```diff
@@ -253,4 +266,38 @@ endif()
+            add_custom_command( PRE_BUILD
+                    OUTPUT ${OUTPUT_FILE}
+                    COMMAND ${Vulkan_slang_EXECUTABLE} ${SHADER_FILE_SLANG} -profile ${SLANG_PROFILE} -target spirv -o ${OUTPUT_FILE} -entry ${SLANG_ENTRY_POINT}
+                    COMMAND ${CMAKE_COMMAND} -E copy ${OUTPUT_FILE} ${directory}
```
You can have multiple entries in one spirv file.  So, instead of having 1 vert shader and 1 frag shader file, you can join them into one slang file and give each an entry point.
This is beneficial especially for ray tracing:
slangc shaders.slang -target spirv -profile spirv_1_4 -emit-spirv-directly -fvk-use-entrypoint-name  -entry vertMain -entry fragMain -o slang.spv
is an example of how to achieve that.

SaschaWillems replied:
Yeah, I am aware of that (and use it in my own projects). But that would've required additional work to our build setup and framework, so I stuck with separate shader files.

SRSaunders commented:
@SaschaWillems I want to point out that iOS is the only CI build that uses the Vulkan SDK, and by extension, dxc and slangc.  That's the reason your "slangc not found" issue (now fixed) showed up only in context of the iOS CI.

For instance, here is what happens on Windows (Linux and macOS are similar):

![Screenshot 2025-03-18 at 8 27 37 PM](https://github.com/user-attachments/assets/211338e1-f460-46e9-9871-cd2001332d16)

As you can see, Vulkan SDK is not found (and dxc + slangc are not used).

And here is what happens on iOS:

![Screenshot 2025-03-18 at 8 29 44 PM](https://github.com/user-attachments/assets/570fb595-7ce0-4f89-919c-457af2b0b4fe)

Vulkan SDK is found and dxc is used.  slangc would be used if the SDK is up-versioned to 1.4.304 (as per my open PR).  

I just want you to be aware of this, in case you want to change the CI for Windows, Linux, macOS to actually use the Vulkan SDK and run the shader compilers (dxc, slangc) as part of the CI build process.  I am less familiar with Android so I won't comment on that.

gpx1000 commented:
@SRSaunders You're almost correct in how our CI works.  All platforms use the Android NDK that is already installed on the runners by github.  If we were to waste time downloading the latest version like we used to when we used the docker build then we'd constantly be playing catch up.  We shovel that responsibility off to github as it stands to reason that most developers would use that version of the SDK.
Now the system installed SDK is ONLY used in CI and the build in general for things in the library, the validation layer or the binary directory on each of the platforms (shader compilers are heavy candidates here).  We have Vulkan headers in our repo so we control what is used for compiling.  This is useful for those that need to make changes to Vulkan library or use a specific version; as well as giving us complete control over the build itself.

Now, iOS is the odd duck out.  The reason why is when I wrote the CI, the SDK wasn't released yet that had iOS libraries in it, there was no way to get it to actually link.  Now that github Apple runners have the version of the SDK which has the iOS libs, it's in my TODO list to one day take out the SDK from being explicitly downloaded and instead rely upon the system runner's version of the SDK for the similar reasons mentioned above for moving away from the docker image files.  The only thing it really effects is what library things get linked against on a CI build that never actually runs.

It is somewhere in the back of my TODO icebox list to have CI actually run samples, and validate that everything works while also demonstrating how to do that with an AI able to check for buttons in the right spot etc.  Eventually I'll get around to it, just haven't gotten there yet.

gpx1000 commented:
er, sorry I meant Vulkan SDK everywhere I say Android, I'm looking into an Android issue right now and have that on my brain.

SaschaWillems commented:
@SRSaunders : If at all we should disable shader compilation on all CI paths. We provide pre-compiled SPV for a reason. But that should be discussed outside of this PR.

asuessenbach reviewed (dismissed):
By the way: why have all those `*.spv` files changed?

SaschaWillems commented:
Argh. That's because HLSL files are always recompiled and I accidentally committed them while adding slang test shaders. Will try to revert those changes :(

SRSaunders commented:
> Now the system installed SDK is ONLY used in CI and the build in general for things in the library, the validation layer or the binary directory on each of the platforms (shader compilers are heavy candidates here). We have Vulkan headers in our repo so we control what is used for compiling.

@gpx1000 I totally agree with this, but I have noticed that iOS is still the _only_ CI build that activates and uses the GitHub cache for the Vulkan SDK, and downloads it as fallback if not found in the cache (note this cache failure is happening all the time now, which is a separate issue that should be looked into).

The concern I have is that none of the other CI builds access the Vulkan SDK at all (cache is not used), and the build logs show no evidence of shader compilation other than for iOS.  Perhaps this is intentional as mentioned by @SaschaWillems above, but I find it strange that you are relying on iOS alone to test the shader build path and ensure sanity.  I would have thought that at least one other platform and config (e.g. Win Rel or Linux Debug) should test for sanity here.  For instance, the only way that @SaschaWillems found the slangc compile issue on earlier SDKs was due to an iOS CI build failure.  It seems like an oversight to me that a least one standard platform is not sanity testing the shader compilation stuff via CI.

SaschaWillems commented:
Can I please ask to move that discussion to a separate issue?

SRSaunders commented:
> Can I please ask to move that discussion to a separate issue?

Yes, I was anticipating that.  I don't have anything further to say, so I will leave it up to @gpx1000 to open a new issue if he sees a problem here and wants to continue.  Thanks.

SaschaWillems committed:
Trying to revert accident HLSL SPIR-V commit

SaschaWillems commented:
> By the way: why have all those `*.spv` files changed?

I have replaced the changed with files with the ones from main, that seemed to have worked.

SaschaWillems committed:
Use SPIRV profile

SaschaWillems committed:
Fix typos in log messages

gary-sweet reviewed (approved):
Doesn't seem to cause any issues for me, though I don't have a slang compiler.
